Improve performance

sklawren opened the pull request:
### Summary of changes
- Reduce number of times warnings are calculated to once per story (previously would calculate 5 times when switching)
- Use [lrt](https://github.com/dfilatov/lrt) to chunk `get100vhWarnings` so it does not block the main thread
- Add "Rescan" button so that component changes can be scanned again (changing the component state, viewport width, etc)

### Details
- Just html state is sufficient for "loading" status
- Lift loading and html state out of Hints component, fixes rules of hooks violation and eliminates expensive re-renders
- Create Loading component
- Only update tab text when warningCount changes
- Separate fast scans from slow scan (currently just `get100vhWarnings`)
- Optimize `get100vhWarnings` and use LRT with a generator to chunk each iteration and make it cancelable

Related: https://github.com/aholachek/storybook-mobile/issues/13

aholachek commented:
This looks awesome, thank you for doing this work! I'll have a review by the end of the day today.

domyen commented:
So stoked for this! Thanks @sklawren and @aholachek 🙇

aholachek commented on src/Hints.js:
```diff
@@ -185,8 +175,19 @@ const Container = styled.div`
+  height: 2.875rem;
+`
+
+const StyledRescanButton = styled.button`
```
```suggestion
const StyledRescanButton = styled.button`
  margin-left: 1rem;
  border-width: 1px;
  border-radius: 3px;
  padding: .2rem .5rem;
  cursor: pointer;
  font-family: inherit;
  color: inherit;
  border: none;
  font-size: 100%;
  background-color: transparent;
  appearance: none;
  box-shadow: none;
  border: 1px solid;
  &:hover {
    background-color: hsla(0, 0%, 0%, 0.1)
  }
`
`
```

aholachek commented on src/register.js:
```diff
@@ -39,9 +39,9 @@ const StateWrapper = ({ children }) => {
 
   useChannel({
-    [STORY_CHANGED]: (...args) => {
+    [STORY_RENDERED]: (...args) => {
```
nice

aholachek reviewed (commented):
Thank you again for all the work you did on this. I really appreciate the investigation you did and the fixes you made, and I would like to merge this ASAP.

My preference would be for removing the 100vh check and all the associated code entirely.  I don't think that one check it is worth the added complexity of the smart scheduling you've added with `lrt`, and would allow us to remove some extra code. How would you feel about removing this in your pr, instead of what you currently have?

You've obviously spent a fair amount of time diagnosing and fixing this problem already, so I would be happy to make a commit to your pr doing the suggested code removal, if you are ok with it.

sklawren commented:
> My preference would be for removing the 100vh check and all the associated code entirely. I don't think that one check it is worth the added complexity of the smart scheduling you've added with `lrt`, and would allow us to remove some extra code. How would you feel about removing this in your pr, instead of what you currently have?

I actually wanted to go the other way and eventually have all of the utilities be managed through the scheduler so that no matter how complex a story is, the addon would never lock up the UI.

Right now, the vh100 check is the least performant, but it's conceivable that in some components, other utilities could be more expensive. `getActiveWarnings` is the 2nd most expensive utility and can take up to 300ms. I was going to address that one next, but I thought it better to do this in steps and first fix the vh100 check.

Also, it's possible that there could be future utilities that would be helpful when developing for mobile and some of them might also be necessarily expensive.

Thoughts?

sklawren commented:
I merged in your Rescan button design change and changed  the hover hsla alpha to 0.15 because 0.1 wasn't quite visible when using the Storybook Dark Mode plugin.

I also changed all the `<a href>` links to open in a blank tab so that you don't leave Storybook when you click them.

I also switched the other expensive operation `getActiveWarnings` to use the scheduler. It's pretty easy. I could do the rest in this PR or another one. Let me know how you'd like to proceed.

sklawren commented:
I have made significant improvements.

- Added prettier to the actual project for IDEs which do not have Prettier built-in.
- Refactored all utilities so they are optimized and iterate as few times as possible (just once in most cases)
- Fixed a few mistakes I found in some utilities
- Improved the scheduling code so that it's much cleaner and simpler to schedule utilities
- Added a spinner to the panel while the scans are running (using the official Storybook one for consistency)
- Only scan when the Addon is active to improve overall performance of Storybook
- Moved the count from the tab into the panel (to go along with the active scan change)
- Updated the text feedback for consistency between states

aholachek reviewed (approved):
Thank you for all the work you've done on this. I've finally had a chance to dig in to the code, and It looks great.
